global: move internal ran functions to global scope

edcdavid opened the pull request:
Note: Now relies on https://github.com/openshift-kni/eco-goinfra/pull/560 to be merged first
Moving some functions previously defined in the RAN scope to global scope:

- ~ExecLocalCommand~
- ExecCmdWithRetries
- ExecCmdWithStdoutWithRetries
- ExecCommandOnSNOWithRetries
- isErrorExecuting
- WaitForClusterRecover
- IsPodHealthy
- IsPodInCondition
- SoftRebootSNO
- waitForClusterReachable
- waitForAllPodsHealthy

This is PR is split from https://github.com/openshift-kni/eco-gotests/pull/83 to review the global code only

kononovn commented on tests/internal/params/vars.go:
```diff
@@ -1,9 +1,17 @@
+	// LogLevel is the verbosity for global tools.
+	LogLevel glog.Level = 90
+	// OpenshiftLoggingNamespace is the namespace where logging pods are.
+	OpenshiftLoggingNamespace = "openshift-logging"
```
This parameter has to be added to https://github.com/openshift-kni/eco-gotests/blob/60d66789c0a6707c226d20d344dc1618ccb368ba/tests/internal/config/config.go#L32 instead of hardcode.

edcdavid replied:
fixed

kononovn commented on tests/internal/params/vars.go:
```diff
@@ -1,9 +1,17 @@
+		"security.openshift.io/scc.podSecurityLabelSync": "false",
+	}
+	// LogLevel is the verbosity for global tools.
+	LogLevel glog.Level = 90
```
It's OK to duplicate this LogLevel no need to move it to global vars.

edcdavid replied:
fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+	ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+	defer cancel()
+
+	glog.V(params.LogLevel).Infof("Locally executing command '%s' with args '%v'", command, args)
```
Please move this log message to the beginning of the function.

edcdavid replied:
fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+		if isErrorExecuting(err) {
+			glog.V(params.LogLevel).Infof("Error during command execution, retry %d (%d max): %w", retry+1, retries, err)
+
+			time.Sleep(10 * time.Second)
```
It'll be good to add sleep interval as a argument.

edcdavid replied:
fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+// second delay between retries, and returns the stdout for each node.
+func ExecCmdWithStdoutWithRetries(
+	client *clients.Settings, retries uint, command string, options ...metav1.ListOptions) (map[string]string, error) {
+	for retry := range retries {
```
Please add glog message

edcdavid replied:
fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+		if isErrorExecuting(err) {
+			glog.V(params.LogLevel).Infof("Error during command execution, retry %d (%d max): %w", retry+1, retries, err)
+
+			time.Sleep(10 * time.Second)
```
It'll be good to add sleep interval as a argument.

edcdavid replied:
fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+// retrying on internal errors retries times
+// with a 10 second delay between retries, and returns the stdout.
+func ExecCommandOnSNOWithRetries(client *clients.Settings, retries uint, command string) (string, error) {
+	outputs, err := ExecCmdWithStdoutWithRetries(client, retries, command)
```
Please add glog message.

edcdavid replied:
fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+}
+
+// isErrorExecuting matches errors that contain the message "error executing command in container".
+func isErrorExecuting(err error) bool {
```
Please move private function to the end of the file.

edcdavid replied:
fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+		return err
+	}
+
+	time.Sleep(extraWait)
```
What is the point of this sleep ?

edcdavid replied:
This was to fix instability after reboot when running commands. This is no longer needed with retries. I removed it.

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+}
+
+// IsPodHealthy returns true if a given pod is healthy, otherwise false.
+func IsPodHealthy(pod *pod.Builder) bool {
```
This function can be removed. Please use eco-goinfra wait functions. You can add additional function is current functionality is not enough.

https://github.com/openshift-kni/eco-goinfra/blob/63369f7a7585eab062658272d3fd1b95b166b770/pkg/pod/pod.go#L274

klaskosk replied:
I agree with Nikita here, there is room for these utilities to be added to eco-goinfra. It's been in the backlog for RAN to reorganize these helpers so they're not as good as they could be

edcdavid replied:
Fixed. Created https://github.com/openshift-kni/eco-goinfra/pull/560

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,209 @@ func ExecCmdWithStdout(
+}
+
+// IsPodInCondition returns true if a given pod is in expected condition, otherwise false.
+func IsPodInCondition(pod *pod.Builder, condition corev1.PodConditionType) bool {
```
Same comment as above: https://github.com/openshift-kni/eco-gotests/pull/128/files#r1698723302

edcdavid replied:
fixed. created https://github.com/openshift-kni/eco-goinfra/pull/560

kononovn reviewed (changes requested):
Please check few comments above.

kononovn commented:
@edcdavid Please change PR name/commit message to something likeL
`global: move internal ran funcs to global scope`

klaskosk commented:
From a RAN perspective, I'm hesitant to move all of these functions out of RAN's directory since as Nikita has noted there's plenty of room for improvement, much of it is specific to the team, and we have to be careful about not breaking CI with changes.

That being said, I'd like to hear @mcornea's thoughts on how system tests would approach these functions in case there's a different way that's more in line with what they're doing. Retries exist because CI would fail when the MCO daemon pod restarted in the middle of a test, so I'm not sure if they have encountered a similar problem.

If moving these functions really is the best way then I'd like to take the time to ensure it doesn't cause any automation issues for RAN but am not otherwise opposed

mcornea commented:
> That being said, I'd like to hear @mcornea's thoughts on how system tests would approach these functions in case there's a different way that's more in line with what they're doing. Retries exist because CI would fail when the MCO daemon pod restarted in the middle of a test, so I'm not sure if they have encountered a similar problem.
> 
IMO we should take a staged approach by first adding these new functions globally and secondly refactor existing tests to use them. Since there's a risk the newly added functions introduce unexpected breakages to existing tests then this change needs to be properly planned according to each team's capacity. I'd personally first add the new global functions, then add the new system tests using the new functions and lastly refactor any of the existing ran or system-tests to make use of the new global functions.

klaskosk commented:
> IMO we should take a staged approach by first adding these new functions globally and secondly refactor existing tests to use them.

This seems like a pretty fair approach. I had already set aside some time next week for the reorganization of ran helpers so I'll be able to dedicate some capacity then to getting these changes tested and implemented.

---

This came up when discussing refactoring inside of ran, but it may be worth considering naming the global package something more like command rather than cluster. Especially if we're moving `ExecLocalCommand` which doesn't do anything cluster related. I'll leave that up to @kononovn however.

edcdavid commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +180,146 @@ func ExecCmdWithStdout(
+
+// ExecCmdWithRetries executes a command on the provided client on each node matching nodeSelector,
+// retrying on internal errors
+// retries times with a 10 second delay between retries, and ignores the stdout.
```
Fixed

edcdavid commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +180,146 @@ func ExecCmdWithStdout(
+
+// ExecCommandOnSNOWithRetries executes a command on the provided single node client,
+// retrying on internal errors retries times
+// with a 10 second delay between retries, and returns the stdout.
```
Fixed.

edcdavid commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +180,146 @@ func ExecCmdWithStdout(
 	return outputMap, nil
 }
+
+// new commands
```
Fixed.

kononovn commented on tests/internal/params/vars.go:
```diff
@@ -1,9 +1,11 @@
-	"pod-security.kubernetes.io/warn":                "privileged",
-	"security.openshift.io/scc.podSecurityLabelSync": "false",
-}
+var (
```
Why did you change it? Can you revert it back to it's original state?

edcdavid replied:
Fixed. I added a variable here before and reformatted.

kononovn commented on tests/internal/config/default.yaml:
```diff
@@ -11,7 +11,7 @@ control_plane_label: "control-plane"
 tc_prefix: "TC-"
 mco_namespace: "openshift-machine-config-operator"
 mco_config_daemon_name: "machine-config-daemon"
+logging_namespace: "openshift-logging"
```
Please re-name to logging_operator_namespace

edcdavid replied:
Fixed

kononovn commented on tests/internal/config/config.go:
```diff
@@ -30,6 +30,7 @@ type GeneralConfig struct {
 	ControlPlaneLabel         string `yaml:"control_plane_label" envconfig:"ECO_CONTROL_PLANE_LABEL"`
 	TCPrefix                  string `yaml:"tc_prefix" envconfig:"ECO_TC_PREFIX"`
 	MCONamespace              string `yaml:"mco_namespace" envconfig:"ECO_MCO_NAMESPACE"`
+	LoggingNamespace          string `yaml:"logging_namespace" envconfig:"ECO_LOGGING_NAMESPACE"`
```
LoggingOperatorNamespace
ECO_LOGGING_OPERATOR_NAMESPACE

edcdavid replied:
Fixed

kononovn reviewed (changes requested):
Please check few comments

klaskosk commented:
This passed the tests for the RAN team so lgtm once Nikita's comments are addressed

edcdavid commented:
@klaskosk I moved the ExecLocalCommand back to RAN area after a comment from @kononovn . I did not recreate the cluster.go, instead I moved it to the ranhelper, since this is running commands locally

kononovn commented on tests/internal/config/config.go:
```diff
@@ -30,6 +30,7 @@ type GeneralConfig struct {
 	ControlPlaneLabel         string `yaml:"control_plane_label" envconfig:"ECO_CONTROL_PLANE_LABEL"`
 	TCPrefix                  string `yaml:"tc_prefix" envconfig:"ECO_TC_PREFIX"`
 	MCONamespace              string `yaml:"mco_namespace" envconfig:"ECO_MCO_NAMESPACE"`
+	LoggingNamespace          string `yaml:"logging_operator_namespace" envconfig:"ECO_LOGGING_OPERATOR_NAMESPACE"`
```
Please change LoggingNamespace > LoggingOperatorNamespace

edcdavid replied:
Fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,130 @@ func ExecCmdWithStdout(
+	glog.V(90).Infof("Executing command '%s' with %d retries and interval %v. Node Selector: %v",
+		command, retries, interval, nodeSelector)
+
+	for retry := range retries {
```
It looks like wait PollImmediate* pattern. Please use wait functions instead of sleep

edcdavid replied:
Fixed

kononovn commented on tests/internal/cluster/cluster.go:
```diff
@@ -171,3 +179,130 @@ func ExecCmdWithStdout(
+		command, retries, interval, options)
+
+	for retry := range retries {
+		outputs, err := ExecCmdWithStdout(client, command, options...)
```
It looks like wait PollImmediate* pattern. Please use wait functions instead of sleep

edcdavid replied:
Fixed

kononovn reviewed (changes requested):
Global level looks good. Just few comments to fix.

kononovn reviewed (approved):
global level 
/lgtm

klaskosk reviewed (approved):
/lgtm for ran
